### PR TITLE
Snap revision prune

### PR DESCRIPTION
**Pull request by seawaywen**
This prune function will remove all the cached revision files expect the latest revision under the snapcraft’s xdg cache(XDG_CACHE_HOME/snapcraft/{project_name}/revisions),

So far I'm uncertain where should this revision prune function be invoked. now it's called right after the revision cache function is triggered in "snap push".

It's the following branch for PR: #889

LP: #1641512

### CODE REVIEW

**Comment by sergiusens** [comment]
Now that #889 is in can we rebase or merge with master?

**Comment by seawaywen** [comment]
@sergiusens my source branch rebased from #889, the conflicts were all resolved, I think the simple merge should work.

File: snapcraft/_store.py [hunk 1]
```diff
@@ -547,6 +548,14 @@ def history(snap_name, series, arch):
     print(tabulated_revisions)
 
 
+def latest_rev(snap_name, series=None, arch=None):
```
Comment (sergiusens) [inline snapcraft/_store.py]: being pythonic and all and this is not a direct reflection of a cli command, can you call this `get_latest_rev` and maybe add a docstring? Also document the return result.

Is there a change for there not to be a latest rev? Maybe some safe guards for that if necessary

File: snapcraft/internal/cache/_snap.py [hunk 2]
```diff
@@ -39,19 +41,43 @@ def cache(self, snap_filename, revision):
...
-        cached_snap_path = os.path.join(snap_cache_dir, cached_snap)
+        cached_snap_path = os.path.join(self.snap_cache_dir, cached_snap)
         try:
             shutil.copyfile(snap_filename, cached_snap_path)
         except OSError:
             logger.warning(
                 'Unable to cache snap {}.'.format(cached_snap))
         return cached_snap_path
 
+    def prune(self, revision=None):
```
Comment (sergiusens) [inline snapcraft/internal/cache/_snap.py]: Can this return a list of what was prunned so the caller can print it?

File: snapcraft/internal/cache/_cache.py [hunk 3]
```diff
@@ -33,13 +33,14 @@ def __init__(self):
     def cache(self):
         raise NotImplementedError
 
-    def prune(self):
+    def prune(self, revision=None):
```
Comment (sergiusens) [inline snapcraft/internal/cache/_cache.py]: `revision` shouldn't be in `SnapcraftCache`, if necessary go with the `**kwargs, *args` trick here for the cases we need to call `super`

**Comment by sergiusens** [comment]
BTW, thanks for the PR

File: snapcraft/_store.py (hunk 1 above)
Comment (seawaywen) [inline snapcraft/_store.py]: as we discussed, remove the *get_lastest_rev* from cache package, we won't have this issue then.

File: snapcraft/internal/cache/_snap.py (hunk 2 above)
Comment (seawaywen) [inline snapcraft/internal/cache/_snap.py]: yes, as you suggested, the function will return the prunned list

File: snapcraft/internal/cache/_cache.py (hunk 3 above)
Comment (seawaywen) [inline snapcraft/internal/cache/_cache.py]: declared in base class like: *def prune(self, *args, **kwargs)*,
and in subclass it looks like: *def prune(self, keep_revision, *args)

**Comment by seawaywen** [comment]
@sergiusens I have updated the branch with some changes as we discussed yesterday. mainly get rid of the `get_latest_revision` in the cache package's business logic.

File: snapcraft/internal/cache/_snap.py [hunk 4]
```diff
@@ -39,19 +40,47 @@ def cache(self, snap_filename, revision):
...
-        cached_snap_path = os.path.join(snap_cache_dir, cached_snap)
+        cached_snap_path = os.path.join(self.snap_cache_dir, cached_snap)
         try:
             shutil.copyfile(snap_filename, cached_snap_path)
         except OSError:
             logger.warning(
                 'Unable to cache snap {}.'.format(cached_snap))
         return cached_snap_path
 
+    def prune(self, keep_revision, *args):
```
Comment (sergiusens) [inline snapcraft/internal/cache/_snap.py]: this should be
```python
def prune(self, *, keep_revision):
```

File: snapcraft/internal/cache/_snap.py [hunk 5]
```diff
@@ -39,19 +40,47 @@ def cache(self, snap_filename, revision):
...
+        """Prune the snap revisions beside the keep_revision in XDG cache.
+
+        :returns: pruned file paths list.
+        """
+        if keep_revision is None:
```
Comment (sergiusens) [inline snapcraft/internal/cache/_snap.py]: This check isn't needed, the one below does the job, if `None` is passed this is what an `int` conversion raises:

```python
int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

File: snapcraft/internal/cache/_snap.py [hunk 6]
```diff
@@ -60,3 +89,22 @@ def _rewrite_snap_filename_with_revision(snap_file, revision):
...
+    split_name_parts = filename.split('_')
+    if len(split_name_parts) != 4:
+        logger.debug('The cached snap filename {} is invalid.'.format(
+            snap_filename))
+        return None
+
+    try:
+        rev = int(split_name_parts[-1])
+    except ValueError:
+        logger.debug('The cached snap filename has invalid revision.')
```
Comment (sergiusens) [inline snapcraft/internal/cache/_snap.py]: typo: has *an* invalid

File: snapcraft/internal/cache/_snap.py [hunk 7]
```diff
@@ -39,19 +40,47 @@ def cache(self, snap_filename, revision):
...
         """
-        snap_cache_dir = self._setup_snap_cache()
-
         cached_snap = _rewrite_snap_filename_with_revision(
             snap_filename,
             revision)
-        cached_snap_path = os.path.join(snap_cache_dir, cached_snap)
+        cached_snap_path = os.path.join(self.snap_cache_dir, cached_snap)
         try:
             shutil.copyfile(snap_filename, cached_snap_path)
```
Comment (sergiusens) [inline snapcraft/internal/cache/_snap.py]: I just noticed this from the previous review; mind changing it to

```python
from snapcraft.file_utils import link_or_copy
...
...
...
try:
    link_or_copy(snap_filename, cached_snap_path)
...
```

File: snapcraft/internal/cache/_snap.py (hunk 4 above)
Comment (seawaywen) [inline snapcraft/internal/cache/_snap.py]: changed the signature as above

File: snapcraft/internal/cache/_snap.py (hunk 5 above)
Comment (seawaywen) [inline snapcraft/internal/cache/_snap.py]: Yes, you're right, thanks pointing out. removed the extra check.

File: snapcraft/internal/cache/_snap.py (hunk 6 above)
Comment (seawaywen) [inline snapcraft/internal/cache/_snap.py]: update the typo

**Pull request merged**